message ReadMessage {
    required string key = 1;
    required uint64 readtime = 2;
    required uint64 end = 3;
}

message ReadReply {
     required string key = 1;
     required string value = 2;
     required uint64 timestamp = 3;
     required uint64 end = 4;
     required int32 op = 5;
}

message WriteMessage {
    required string key = 1;
    required string value = 2;
}

message IncrementMessage {
   required string key = 1;
   required int32 inc = 2;
}

// A serialized transaction
message TransactionMessage {
    required int32 mode = 1;
    required uint64 timestamp = 2;
    repeated ReadMessage readset = 3;
    repeated WriteMessage writeset = 4;
    repeated IncrementMessage incrementset = 5;
}
